include/sentry.h

```diff
@@ -477,32 +477,60 @@ SENTRY_API int sentry_envelope_write_to_file(
     const sentry_envelope_t *envelope, const char *path);
 
 /**
- * Type of the callback for transports.
+ * This represents an interface for user-defined transports.
+ *
+ * Transports are responsible for sending envelopes to sentry and are the last
+ * step in the event pipeline. A transport has the following hooks:
+ *
+ * * `send_func`: This function will take ownership of an envelope, and is
+ *   responsible for freeing it via `sentry_envelope_free`.
+ * * `startup_func`: This hook will be called by sentry and instructs the
+ *   transport to initialize itself.
+ * * `shutdown_func`: Instructs the transport to flush its queue and shut down.
+ *   This hook receives a millisecond-resolution `timeout` parameter and should
+ *   return `true` when the transport was flushed and shut down successfully.
+ *   In case of `false`, sentry will log an error, but continue with freeing the
+ *   transport.
+ * * `free_func`: Frees the provided `data`. This hook might be called even
+ *   though `shudown_func` returned `false` previously.
+ *
+ * The transport interface might be extended in the future with hooks to flush
+ * its internal queue without shutting down, and to dump its internal queue to
+ * disk in case of a crash.
  */
-typedef void (*sentry_transport_function_t)(
-    const sentry_envelope_t *envelope, void *data);
+struct sentry_transport_s;
+typedef struct sentry_transport_s sentry_transport_t;
 
 /**
- * This represents an interface for user-defined transports.
+ * Creates a new transport.
  *
- * This type is *deprecated*, and will be replaced by an opaque pointer type and
- * builder methods in a future version.
+ * This sets both the send hook, as well as the user-provided `data`.
```

Comment:
It's probably worth mentioning that if `data` is passed, a free func **must** be provided. This gets me wondering if we should move this, instead:

- `new`: only create a transport without data.
- `set_data` or probably better `set_state`: set the state and provide a free func in one go.

There may of course be an edge case when the data is static and doesn't need to be freed, but in this case passing an explicit `NULL` to the free func sounds much more defensive.

Reply:
I was actually thinking about pairing up functions, but I think its better this way. This way the function signature is also the same as with the previous function transport. I will add a comment about free.

Reply:
Since we're breaking anyway, I think it's worth reconsidering this. My fear is that with the current signature (especially when migrating!) it is way too easy to produce a memory leak.